lightning/src/routing/scoring.rs

```diff
@@ -460,25 +458,466 @@ impl<T: Time> Readable for ChannelFailure<T> {
+	fn as_directed_mut(
+		&mut self, source: &NodeId, target: &NodeId, capacity_msat: u64, half_life: Duration
+	) -> DirectedChannelLiquidity<&mut u64, T, &mut T> {
+		let (min_liquidity_offset_msat, max_liquidity_offset_msat) = if source < target {
+			(&mut self.min_liquidity_offset_msat, &mut self.max_liquidity_offset_msat)
+		} else {
+			(&mut self.max_liquidity_offset_msat, &mut self.min_liquidity_offset_msat)
+		};
+
+		DirectedChannelLiquidity {
+			min_liquidity_offset_msat,
+			max_liquidity_offset_msat,
+			capacity_msat,
+			last_updated: &mut self.last_updated,
+			now: T::now(),
+			half_life,
+		}
+	}
+}
+
+impl<L: Deref<Target = u64>, T: Time, U: Deref<Target = T>> DirectedChannelLiquidity<L, T, U> {
+	/// Returns the success probability of routing the given HTLC `amount_msat` through the channel
+	/// in this direction.
+	fn success_probability(&self, amount_msat: u64) -> Probability {
+		let max_liquidity_msat = self.max_liquidity_msat();
+		let min_liquidity_msat = core::cmp::min(self.min_liquidity_msat(), max_liquidity_msat);
+		if amount_msat > max_liquidity_msat {
+			Probability::Zero
+		} else if amount_msat < min_liquidity_msat {
+			Probability::One
+		} else {
+			let numerator = max_liquidity_msat + 1 - amount_msat;
+			let denominator = max_liquidity_msat + 1 - min_liquidity_msat;
+			if numerator == denominator {
+				Probability::One
+			} else {
+				Probability::Ratio { numerator, denominator }
+			}
+		}
+	}
+
+	/// Returns the lower bound of the channel liquidity balance in this direction.
+	fn min_liquidity_msat(&self) -> u64 {
+		self.decayed_offset_msat(*self.min_liquidity_offset_msat)
+	}
+
+	/// Returns the upper bound of the channel liquidity balance in this direction.
+	fn max_liquidity_msat(&self) -> u64 {
+		self.capacity_msat
+			.checked_sub(self.decayed_offset_msat(*self.max_liquidity_offset_msat))
+			.unwrap_or(0)
+	}
+
+	fn decayed_offset_msat(&self, offset_msat: u64) -> u64 {
+		self.now.duration_since(*self.last_updated).as_secs()
+			.checked_div(self.half_life.as_secs())
+			.and_then(|decays| offset_msat.checked_shr(decays as u32))
+			.unwrap_or(0)
+	}
+}
+
+impl<L: DerefMut<Target = u64>, T: Time, U: DerefMut<Target = T>> DirectedChannelLiquidity<L, T, U> {
+	/// Adjusts the channel liquidity balance bounds when failing to route `amount_msat`.
+	fn failed_at_channel(&mut self, amount_msat: u64) {
+		if amount_msat < self.max_liquidity_msat() {
+			self.set_max_liquidity_msat(amount_msat);
+		}
+	}
+
+	/// Adjusts the channel liquidity balance bounds when failing to route `amount_msat` downstream.
+	fn failed_downstream(&mut self, amount_msat: u64) {
+		if amount_msat > self.min_liquidity_msat() {
+			self.set_min_liquidity_msat(amount_msat);
+		}
+	}
+
+	/// Adjusts the channel liquidity balance bounds when successfully routing `amount_msat`.
+	fn successful(&mut self, amount_msat: u64) {
+		let max_liquidity_msat = self.max_liquidity_msat().checked_sub(amount_msat).unwrap_or(0);
```

Comment:
I am a but confused if this and the prior two functions always work properly with concurrency. But it seems to me that concurrency is actually not  messing up the book keeping here. But how do you handle the decay of the knowledge when substracting from max_liquidity? This could happen at two very different moments. Let's say you forward one htlc at 12:00 and the next at 12:45 both have been successful. How will you compute the max liquidity at 13:05? As far as I understand the knowledge from the first routing event should have been decyed to half at that time. I am not sure how this is reflected in this data structure. might not be a problem as the halflife is chosen arbitrarlity at this time anyway but I thought I should point this out.